### PR TITLE
Advertise multiple listeners via Alt-Svc and improve perf of SetQuicHeaders

### PR DESCRIPTION
**Pull request by renbou**
PR to resolve https://github.com/lucas-clemente/quic-go/issues/3350.
Looking at the profile of Caddy with this commit used, `SetQuicHeaders` seems to work 5 times faster now, and pretty much consists of a single slice append.

image

### CODE REVIEW

**Review by marten-seemann — COMMENTED** [review]
There is a valid use case for using a listener running on top of a `net.PacketConn` that is not a `net.UDPConn`, and which might not have a port to begin with. I'm wondering if we should just log the error, and skip addresses for which we can't determine the port. We could then extract the port on `addListener`.
wdyt?

File: http3/server.go [hunk 1]
```diff
@@ -237,21 +239,78 @@ func (s *Server) serveImpl(startListener func() (quic.EarlyListener, error)) err
...
+		}
+	}
+
+	extractPort := func(addr net.Addr) (int, error) {
+		_, portStr, err := net.SplitHostPort(addr.String())
+		if err != nil {
+			return 0, err
+		}
+
+		portInt, err := net.LookupPort("tcp", portStr)
```
[the remark this answers is not in the document]
Comment (renbou) [reply http3/server.go:263]: Well, there's technically a test which supplies ":https" as the port, and it is valid to run `http3.Server.ListenAndServe` with an address like ":https" given (which should launch a udp listener on 0.0.0.0:443). I don't really find it meaningful to support this and would rather have the function simply split by ':' and then do an `Atoi`, so that `SetQuicHeaders` doesn't return an error. But since it's already supported and `SetQuicHeaders` returns an error, doesn't make sense to change the API now :confused:

**Comment by renbou** [comment]
> There is a valid use case for using a listener running on top of a `net.PacketConn` that is not a `net.UDPConn`, and which might not have a port to begin with. I'm wondering if we should just log the error, and skip addresses for which we can't determine the port. We could then extract the port on `addListener`. wdyt?

What do you mean by "extract the port on `addListener`"? We're already extracting the port during `addListener`. Hmm, I could technically see someone using the http3 server on a unix socket, I think in the case where an address doesn't have a port, we can use `Server.Addr` (and if `Server.Port` then that would be used instead by default). And if `Server.Addr` isn't set then we should probably log the error and return it during `SetQuicHeaders`, yeah. I don't think advertising 443 by default is a good idea since 443 only really makes sense in terms of http1.1/2 over tls, but http3 can run on any port.

**Comment by marten-seemann** [comment]
I wasn't suggesting to use 443 by default. Sorry for the confusion.

Let me rephrase: We could get rid of the `altSvcGenerationErr`. When an error occurs in `generateAltSvcHeader`, we just log that error (even better: we only log it once, by running the port extraction logic in `addListener`), and skip that listener. That would make it possible to use (some) listeners that don't have port numbers, and still use `SetQUICHeaders`.

**Comment by renbou** [comment]
Ah, okay, this makes sense now. Though I think in this case, if no other ports are found from the listeners, I think we should try using `Server.Addr` as a last resort, and if it fails, then we could set an error and return it in `SetQUICHeaders`. I don't think there's a way to make `SetQUICHeaders` work nicely without returning an error at all, and that would also change existing API. Though now that I think about it, since in this case we would only return an error when no port is available for us to return valid headers with in `SetQUICHeaders`, we can still remove `altSvcGenerationErr` and instead return an error specifying that no ports are available (when `altSvcHeader == ""`, for example). I'll implement this so it's easier to understand what I mean

**Review by marten-seemann — APPROVED** [review]

**Comment by marten-seemann** [comment]
Thank you!
